dartagnan/src/main/java/com/dat3m/dartagnan/analysis/Refinement.java

```diff
@@ -0,0 +1,387 @@
+        Program program = task.getProgram();
+        GraphRefinement refinement = new GraphRefinement(task);
+        Result res = UNKNOWN;
+
+        // ====== Test code ======
+        List<Function<Event, Event>> perms = computePerms(task); // Used for symmetry learning
+        // ----------
+        if (ENABLE_SYMMETRY_BREAKING) {
+            prover.addConstraint(task.encodeSymmetryBreaking(ctx));
+        }
+        // =======================
+
+        prover.push();
+        prover.addConstraint(task.encodeAssertions(ctx));
+
+        // Just for statistics
+        List<DNF<CoreLiteral>> foundViolations = new ArrayList<>();
+        List<RefinementStats> statList = new ArrayList<>();
+        int vioCount = 0;
+        long lastTime = System.currentTimeMillis();
+        long curTime;
+        long totalSolvingTime = 0;
+
+        while (!prover.isUnsat()) {
+            curTime = System.currentTimeMillis();
+            totalSolvingTime += (curTime - lastTime);
+
+            if (REF_PRINT_STATISTICS) {
+                System.out.println(" ===== Iteration: " + ++vioCount + " =====");
+                System.out.println("Solving time( ms): " + (curTime - lastTime));
+            }
+
+            RefinementResult gRes;
+            try (Model model = prover.getModel()) {
+                gRes = refinement.kSearch(model, ctx, 2);
+            }
+            RefinementStats stats = gRes.getStatistics();
+            statList.add(stats);
+            if (REF_PRINT_STATISTICS) {
+                System.out.println(stats);
+            }
+
+            res = gRes.getResult();
+            if (res == FAIL) {
+                DNF<CoreLiteral> violations = gRes.getViolations();
+                foundViolations.add(violations);
+                refine(prover, ctx, violations, perms);
+                // Some statistics
+                if (REF_PRINT_STATISTICS) {
+                    for (Conjunction<CoreLiteral> cube : violations.getCubes()) {
+                        System.out.println("Violation size: " + cube.getSize());
+                        System.out.println(cube);
+                    }
+                }
+            } else {
+                // No violations found, we can't refine
+                break;
+            }
+            lastTime = System.currentTimeMillis();
+        }
+        curTime = System.currentTimeMillis();
+        totalSolvingTime += (curTime - lastTime);
+        if (REF_PRINT_STATISTICS) {
+            System.out.println(" ===== Final Iteration: " + (vioCount + 1) + " =====");
+            System.out.println("Solving/Proof time(ms): " + (curTime - lastTime));
+        }
+
+        // Possible outcomes: - check() == SAT && res == UNKNOWN -> Inconclusive
+        //                    - check() == SAT && res == PASS -> Unsafe
+        //                    - check() == UNSAT -> Safe
+
+
+        boolean isSat = !prover.isUnsat();
+        if (REF_PRINT_STATISTICS) {
```

Comment:
I don't think this is true. Res=FAIL can never hold, cause it will loop in this case. Also Res=PASS is returned by kSearch, if the violation is actually consistent. So the meaning is PASS=Consistent, FAIL=Inconsistent, UNKNOWN=Inconclusive. I think I will add a new enum for these cases and not use reuse FAIL/PASS/UNKNOWN, so it is more clear.